storage/database/sharded_database.go

```diff
@@ -155,75 +160,244 @@ func (db *shardedDB) Close() {
+		return false
+	}
+	it.key, it.value = e.Key, e.Val
 	return true
 }
 
-func (pdi *shardedDBIterator) Error() error {
-	// TODO-Klaytn implement this later.
+func (it *shardedDBIterator) Error() error {
+	for i, iter := range it.iterators {
+		if iter.Error() != nil {
+			logger.Error("[shardedDBIterator] error from iterator",
+				"err", iter.Error(), "shardNum", i, "key", it.key, "val", it.value)
+			return iter.Error()
+		}
+	}
 	return nil
 }
 
-func (pdi *shardedDBIterator) Key() []byte {
-	// TODO-Klaytn implement this later.
-	return nil
+func (it *shardedDBIterator) Key() []byte {
+	return it.key
 }
 
-func (pdi *shardedDBIterator) Value() []byte {
-	// TODO-Klaytn implement this later.
-	return nil
+func (it *shardedDBIterator) Value() []byte {
+	return it.value
+}
+
+// shardedDBIteratorUnsorted iterates all items of each shardDB.
+// This is useful when you want to get items fast in serial.
+type shardedDBIteratorUnsorted struct {
+	shardedDBIterator
+}
+
+// NewIteratorUnsorted creates a iterator over the entire keyspace contained within
+// the key-value database.
+// If you want to get ordered items in serial, checkout shardedDB.NewIterator()
+// If you want to get items in parallel from channels, checkout shardedDB.NewChanIterator()
+func (db *shardedDB) NewIteratorUnsorted(prefix []byte, start []byte) Iterator {
+	resultCh := make(chan common.Entry, shardedDBCombineChanSize)
+	it := &shardedDBIteratorUnsorted{
+		shardedDBIterator{
+			shardedDBChanIterator: db.NewChanIterator(context.TODO(), prefix, start, resultCh),
+			resultCh:              resultCh,
+		}}
+	return it
+}
+
+// shardedDBChanIterator creates iterators for each shard DB.
+// Channels subscribing each iterators can be gained.
+// Each iterators fetch values in binary-alphabetical order.
+// This is useful when you want to operate on each items in parallel.
+type shardedDBChanIterator struct {
+	ctx    context.Context
+	cancel context.CancelFunc
+
+	iterators []Iterator
+
+	combinedChan bool // all workers put items to one resultChan
+	shardNum     int  // num of shards left to iterate
+	resultChs    []chan common.Entry
+}
+
+// NewChanIterator creates iterators for each shard DB. This is useful when you
+// want to operate on each items in parallel.
+// If `resultCh` is given, all items are written to `resultCh`, unsorted with a
+// particular key prefix, starting at a particular initial key. If `resultCh`
+// is not given, new channels are created for each DB. Items are written to
+// corresponding channels in binary-alphabetical order. The channels can be
+// gained by calling `Channels()`.
+//
+// If you want to get ordered items in serial, checkout shardedDB.NewIterator()
+// If you want to get unordered items in serial with Iterator Interface,
+// checkout shardedDB.NewIteratorUnsorted().
+func (db *shardedDB) NewChanIterator(ctx context.Context, prefix []byte, start []byte, resultCh chan common.Entry) shardedDBChanIterator {
```

Comment:
I was planning to use this function from the DB migration code.
If we are using `NewIterator` or `NewIteratorUnsorted`, all data goes through one combined channel. This will makes a bottleneck in performance. So, I left `NewChanIterator` for parallel data iteration.

Reply:
I understood. Thanks. I cound't figure out the difference `NewIteratorUnsorted` and `NewChanIterator`.
As you menthioned in the previous thread, it will help to understand.